File: aoc2015/11.rb
password = "hxbxwxba"

class String
  def password_ok?
    return false if self =~ /[iol]/
    return false unless chars.each_cons(3).any? { |a,b,c| a.ord == b.ord-1 && b.ord == c.ord-1 }
    return false unless self =~ /(.)\1.*(.)\2/
    true
  end

  def next_password
    s = dup.next
    s.next! until s.password_ok?
    s
  end
end

# p "hijklmmn".password_ok?
# p "abbceffg".password_ok?
# p "abbcegjk".password_ok?
# puts
# p "abcdefgh".next_password # abcdffaa
# p "ghijklmn".next_password # ghjaabcc

part1 = password.next_password
part2 = part1.next_password

puts part1, part2

File: aoc2015/haskell/11.hs
import Data.Char (ord, chr)

straight (a : b : c : s)
  | (ord a) == (ord b) - 1 && (ord b) == (ord c) -1 = True
  | otherwise                                       = straight (b : c : s)
straight _ = False

confusing (c : s) = elem c "iol" || confusing s
confusing _ = False

pairs n (a : b : s)
  | a == b    = pairs (n - 1) s
  | otherwise = pairs n (b : s)
pairs n _     = n == 0

passwordOk s = straight s && not (confusing s) && pairs 2 s

nextString' []          = ['a']
nextString' ('z':xs)    = 'a' : nextString' xs
nextString' (x:xs)      = succ x : xs
nextString = reverse . nextString' . reverse

nextPassword s = let next = nextString s in
                 if passwordOk next then next
                 else nextPassword next


main = do
  let password = "hxbxwxba"
      part1 = nextPassword password
      part2 = nextPassword part1
  putStrLn part1
  putStrLn part2
